libcudacxx/include/cuda/__mdspan/host_device_accessor.h

```diff
@@ -0,0 +1,383 @@
+  static constexpr bool __is_offset_noexcept = noexcept(_Accessor{}.offset(data_handle_type{}, 0));
+
+  template <typename data_handle_type>
+  _LIBCUDACXX_HIDE_FROM_ABI static constexpr bool __is_host_accessible_pointer(data_handle_type __p) noexcept
+  {
+    if constexpr (_CUDA_VSTD::is_pointer_v<data_handle_type>)
+    {
+      ::cudaPointerAttributes __attrib{};
+      _CCCL_VERIFY(::cudaPointerGetAttributes(&__attrib, __p) == ::cudaSuccess, "cudaPointerGetAttributes failed");
+      return __attrib.hostPointer != nullptr || __attrib.type == ::cudaMemoryTypeUnregistered;
+    }
+    else
+    {
+      return true; // cannot be verified
+    }
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI static constexpr void __check_host_pointer(data_handle_type __p) noexcept
+  {
+    _CCCL_ASSERT(__is_host_accessible_pointer(__p), "cuda::__host_accessor data handle is not a HOST pointer");
+  }
+
+public:
+  _CCCL_TEMPLATE(typename _NotUsed = void)
+  _CCCL_REQUIRES(_CCCL_TRAIT(_CUDA_VSTD::is_default_constructible, _Accessor))
+  _LIBCUDACXX_HIDE_FROM_ABI __host_accessor() noexcept(_CUDA_VSTD::is_nothrow_default_constructible_v<_Accessor>)
+      : _Accessor{}
+  {}
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr __host_accessor(const _Accessor& __acc) //
+    noexcept(_CUDA_VSTD::is_nothrow_copy_constructible_v<_Accessor>)
+      : _Accessor{__acc}
+  {}
+
+  template <typename _OtherAccessor>
+  __host_accessor(const __device_accessor<_OtherAccessor>&) = delete;
+
+  _CCCL_TEMPLATE(typename _OtherAccessor)
+  _CCCL_REQUIRES(_CCCL_TRAIT(_CUDA_VSTD::is_constructible, _OtherAccessor))
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr
+#if _CCCL_STD_VER >= 2020
+    explicit(!_CUDA_VSTD::is_convertible_v<_OtherAccessor, _Accessor>)
+#endif
+      __host_accessor(const __host_accessor<_OtherAccessor>& __acc) //
+    noexcept(noexcept(_Accessor{_CUDA_VSTD::declval<_OtherAccessor>()}))
+      : _Accessor{__acc}
+  {}
+
+  _CCCL_TEMPLATE(typename _OtherAccessor)
+  _CCCL_REQUIRES(_CCCL_TRAIT(_CUDA_VSTD::is_constructible, _OtherAccessor))
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr
+#if _CCCL_STD_VER >= 2020
+    explicit(!_CUDA_VSTD::is_convertible_v<_OtherAccessor, _Accessor>)
+#endif
+      __host_accessor(const __managed_accessor<_OtherAccessor>& __acc) //
+    noexcept(noexcept(_Accessor{_CUDA_VSTD::declval<_OtherAccessor>()}))
+      : _Accessor{__acc}
+  {}
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr reference access(data_handle_type __p, size_t __i) const
+    noexcept(__is_access_noexcept)
+  {
+    NV_IF_ELSE_TARGET(NV_IS_HOST,
+                      (__check_host_pointer(__p);), //
+                      (static_assert(false, "cuda::__host_accessor cannot be used in DEVICE code");))
+    return _Accessor::access(__p, __i);
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr data_handle_type offset(data_handle_type __p, size_t __i) const
+    noexcept(__is_offset_noexcept)
+  {
+    return _Accessor::offset(__p, __i);
```

Comment:
Question: I am wondering whether we also want to guard this based on host / device